enikshay adherence update task

sravfeyn opened the pull request:
(mainly pseudo-code for review)

sravfeyn committed:
WIP; adherence update task

dimagimon commented on custom/enikshay/case_utils.py:
```diff
@@ -111,6 +111,10 @@ def get_open_episode_case_from_person(domain, person_case_id):
+    return get_adherence_cases_between_dates_from_episode(domain, episode, start_date, end_date)
+
+
+def get_adherence_cases_between_dates_from_episode(domain, episode, start_date, end_date)
```
E901 SyntaxError: invalid syntax

dimagimon commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,88 @@
+            # calculate 'aggregated_score_count_taken'
+            adherence_cases = get_adherence_cases_between_dates_from_episode(
+                DOMAIN
+                episode,
```
E901 SyntaxError: invalid syntax

dimagimon commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,88 @@
+            # calculate 'expected' score
+            adherence_schedule_id = episode.adherence_schedule_id or DAILY_SCHEDULE_ID
+            doses_per_week = DOESES_BY_SCHEDULE_BY_ID[adherence_schedule_id]
+            episode.expected = ((aggregated_score_date_calculated - adherence_schedule_date_start) / 7) * doses_per_week
```
E501 line too long (120 > 115 characters)

czue commented:
@sravfeyn let me know if you want this reviewed

sravfeyn committed:
simple updates

sravfeyn commented:
Yes, please! I am looking for adherence logic review to make sure that it is spec-correct. I will try to restructure the code, to make it easy for tests and will write some tests. So I am not looking for review on that yet.

@proteusvacuum @ctsims

dimagimon commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+        by_date[adherance_date].append(case)
+    return by_date
+
+def get_adherence_schedule_date_start(episode_case):
```
E302 expected 2 blank lines, found 1

dimagimon replied:
E302 expected 2 blank lines, found 1

dimagimon commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+            adherence_schedule_id = episode.adherence_schedule_id or DAILY_SCHEDULE_ID
+            # Todo: what if 'adherence_schedule_id' doesn't exist in fixtures
+            doses_per_week = dose_data.get(adherence_schedule_id)
+            episode.expected = ((aggregated_score_date_calculated - adherence_schedule_date_start) / 7) * doses_per_week
```
E501 line too long (120 > 115 characters)

proteusvacuum replied:
I think you should call this property `expected_doses_taken` or something more descriptive (also elsewhere)

sravfeyn replied:
Okay, I have got it from the spec (assumed the property was so named in the app). Should we update the spec with that title for record? I can update the code

proteusvacuum replied:
I think this should actually be `aggregated_score_count_expected`. That's what is used in the app.
@ctsims

dimagimon replied:
E501 line too long (120 > 115 characters)

sravfeyn commented:
@ctsims @czue @proteusvacuum Bumping this to get [spec signoff](https://docs.google.com/document/d/1FjSdLYOYUCRBuW3aSxvu3Z5kvcN6JKbpDFDToCgead8/edit?pli=1#heading=h.ldp1ikmaxcg)

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+        adherence_schedule_date_start = get_adherence_schedule_date_start(episode)
+        # Todo: What if adherence_schedule_date_start is None
+        if adherence_schedule_date_start > PURGE_DATE:
+            episode.aggregate_date = adherence_schedule_date_start - 1
```
I think you will have to submit a form against the episode to update it (but you probably already knew that)... and I think it might be good to chunk those submissions into batches.

proteusvacuum replied:
I *think* this should be `aggregated_score_date_calculated`. @ctsims to confirm

sravfeyn replied:
Yeah, something like `submit_case_blocks`. Thanks

sravfeyn replied:
👍

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+    return case_accessor.iter_cases(case_ids)
+
+
+def get_latest_adherence_case_for_episode(episode):
```
👍

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+DOSE_TAKEN_INDICATORS = [
+    'directly_observed_dose',
+    'unobserved_dose',
+    'self_administered_dos',
```
`self_administered_dose`

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+        if adherence_schedule_date_start > PURGE_DATE:
+            episode.aggregate_date = adherence_schedule_date_start - 1
+            episode.expected = 0
+            episode.taken = 0
```
I think this should be `aggregated_score_count_taken`

sravfeyn replied:
👍

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+
+def get_doses_data():
+    # return 'doses_per_week' by 'schedule_id' from the Fixture data
+    fixtures = FixtureDataItem.get_indexed_items(DOMAIN, DAILY_SCHEDULE_ID, 'schedule_id')
```
The lookup table is called `adherence_schedules`, and the field is called `id`: https://india.commcarehq.org/a/enikshay/fixtures/view_lookup_tables/?table_id=1ba69af232d61c7479fc6fdecb94cc7e

proteusvacuum reviewed (commented):
Looks like a good start sravs. There was only a few things that I clarified.

sravfeyn commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+    return dict((k, int(fixture['doses_per_week'])) for k, fixture in fixtures.items())
+
+
+def get_open_episode_cases():
```
Sorry, I meant to do `get_open_case_ids_in_domain_by_type(CASE_TYPE_EPISODE)`. Not by `CASE_TYPE_PERSON`.

sravfeyn commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+
+def update_adherence_properties():
+    # edge cases around datetime
+    PURGE_DATE = datetime.datetime.today() - datetime.timedelta(days=60)
```
What kind of verification? By `purge_date`, I understood as date 60 days ago. Please can you list any more spec details as well?

proteusvacuum replied:
Verify with the @ctsims that 60 days is the appropriate time window to do the purge.

sravfeyn committed:
spec update

sravfeyn commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+            episode.taken = 0
+        else:
+            adherence_case = get_latest_adherence_case_for_episode(episode)
+            # ToDo: what if adherence_case doesn't exist?
```
@proteusvacuum, I have added some 'edge-case' questions titled 'Todo', can you respond on what should happen in such cases.

proteusvacuum replied:
I think the `aggregated_score_count_taken` will be 0, and the `aggregated_score_date_calculated` will be the `PURGE_DATE`

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+    'self_administered_dos',
+]
+DAILY_SCHEDULE_ID = 'schedule_daily'
+CASE_TYPE_PERSON = 'person'
```
FYI, all the case type constants are defined in `case_utils.py`

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+from corehq.apps.fixtures.models import FixtureDataItem
+
+
+DOMAIN = 'enikshay'
```
This seems pretty brittle, and would be good to be able to use on our `enikshay-test` domain too. Maybe make this a feature flag thing?

sravfeyn replied:
Yeah. I want to look at flexibility of this. Just wanted to get spec sign off for now

proteusvacuum replied:
cool. I'm signed off. Let me know if you have any more questions

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+           case.type == CASE_TYPE_ADHERENCE and
+           adherence_date > latest_date):
+            latest_date = adherence_date
+            latest_case = adherence_date
```
I think this should be `latest_case = case`

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+            # calculate 'expected' score
+            dose_data = get_doses_data()
+            adherence_schedule_id = episode.adherence_schedule_id or DAILY_SCHEDULE_ID
+            # Todo: what if 'adherence_schedule_id' doesn't exist in fixtures
```
Probably just set `expected_doses_taken` to 0 and raise a soft_assert.

proteusvacuum commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+
+    for episode in get_open_episode_cases():
+        adherence_schedule_date_start = get_adherence_schedule_date_start(episode)
+        # Todo: What if adherence_schedule_date_start is None
```
I don't think these calculations should be run in that case because I think that would indicated that an adherence schedule hasn't been selected. @ctsims to confirm.

sravfeyn replied:
👍

sravfeyn committed:
Merge branch 'master' of github.com:dimagi/commcare-hq into sr-uatbc

sravfeyn committed:
rewrite task differently

sravfeyn committed:
minor changes

sravfeyn committed:
tests; WIP

sravfeyn committed:
Merge branch 'master' into sr-uatbc

sravfeyn committed:
refactor to be modular

sravfeyn committed:
real testing

sravfeyn committed:
mark todo

dimagimon commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,109 @@
+            # calculate 'aggregated_score_count_taken'
+            adherence_cases = get_adherence_cases_between_dates_from_episode(
+                DOMAIN
+                episode,
```
E901 SyntaxError: invalid syntax

sravfeyn committed:
minor changes

sravfeyn committed:
Merge master; resolve conflicts

sravfeyn committed:
Remove prints

sravfeyn committed:
util method for toggles

sravfeyn committed:
get actual domains

sravfeyn commented:
@proteusvacuum @czue This is ready for final review. I am removing `don't merge` label

sravfeyn committed:
update test case

sravfeyn commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,292 @@
+
+        return EpisodeUpdate(episode, self.case_updater)
+
+    def test_adherence_schedule_date_start_late(self):
```
@ctsims 
Hey Clayton, these are some test-cases for adherence update. Can you review these test cases and make sure they are upto spec?

czue replied:
these look like great tests. @ctsims can you please check the logic and confirm it looks right?

sravfeyn committed:
Remove line

sravfeyn committed:
properly add case blocks

dimagimon commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+        if update:
+            return CaseBlock(**{
+                'case_id': self.episode.case_id
+                'create': False,
```
E901 SyntaxError: invalid syntax

sravfeyn committed:
comma

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+        updater.run()
+
+
+class PeriodicCaseUpdater(object):
```
could be nice to add a few more doc strings / comments to this code. is a bit hard to wrap your head around what it's doing without having a good amount of familiarity with the project needs

czue replied:
i also think linking to the spec doc somewhere in this file would be helpful

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+        latest_date = pytz.UTC.localize(datetime.datetime(1990, 1, 1))
+        latest_case = None
+        for case in indexed_cases:
+            adherence_date = parse_datetime(case.dynamic_case_properties().get('adherence_date'))
```
does this crash if `adherence_date` is missing or malformed? if so, might want to account for that since my current read is that the code as-is would crash and nothing in the entire domain would get updated.

sravfeyn replied:
I haven't accounted for exceptions, will add try-catches wherever needed

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+        self.purge_date = pytz.UTC.localize(datetime.datetime.today() - datetime.timedelta(days=60))
+
+    def run(self):
+        for episode in self._get_open_episode_cases():
```
what do you think about putting this whole loop in a try/catch/notify block so individual errors don't cause the entire domain's data to fail?

sravfeyn replied:
Great point, will add it.

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+            with latest 'adherence_date' property
+        """
+        case_accessor = CaseAccessors(self.case_updater.domain)
+        indexed_cases = case_accessor.get_reverse_indexed_cases([self.episode.case_id])
```
it's pretty tempting to try and push this logic into elasticsearch. as-is it seems like this will have to load every single adherence case for every single episode in the entire domain every time it runs right? I think that is going to be pretty untenable long term. could we possibly use the case search index for this?

sravfeyn replied:
I will chat offline with you re. this

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+
+    def __init__(self, domain):
+        self.domain = domain
+        self.purge_date = pytz.UTC.localize(datetime.datetime.today() - datetime.timedelta(days=60))
```
the intention of this is to define the window in which we don't need to worry about running this task based on your schedule start date right? i'm trying to wrap my head around whether this is ok. is there a specific part of the spec you can refer me to? it feels wrong to me that we don't do anything to reconcile your first 60 days of data.

sravfeyn replied:
It's mentioned here
https://docs.google.com/a/dimagi.com/document/d/1FjSdLYOYUCRBuW3aSxvu3Z5kvcN6JKbpDFDToCgead8/edit?pli=1&disco=AAAAAv94gJg

I don't honestly understand the project context of purge date, I have tried to follow the spec and write good test cases to validate

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+                'aggregated_score_date_calculated': adherence_schedule_date_start - datetime.timedelta(days=1),
+                'expected_doses_taken': 0,
+                'aggregated_score_count_taken': 0
+            }
```
an api like this feels like a pretty good place to use a real class or a `namedtuple`

czue replied:
(although I think i realized later that these are case properties that get passed to `update` so maybe not worth it)

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,180 @@
+
+            # calculate 'aggregated_score_count_taken'
+            if adherence_case:
+                adherence_cases = get_adherence_cases_between_dates_from_episode(
```
if we are going to load all the adherence cases from the episode to get the latest one, we should at least reuse that list here to not have to go back to the DB. Again, moving to elasticsearch would probably help a lot for performance of this.

czue commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,291 @@
+        self.factory = CaseFactory(domain=self.domain)
+        self.user = CommCareUser.create(
+            self.domain,
+            "jon-snow@user",
```
nice to see you sticking with enikshay testing conventions. :wolf:

czue commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,291 @@
+
+    def setUp(self):
+        super(TestAdherenceUpdater, self).setUp()
+        delete_all_users()
```
is this definitely necessary?

sravfeyn replied:
I was going to remove, had to put it there for some cleanup.

czue commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,291 @@
+            cls.data_items.append(data_item)
+
+    def tearDown(self):
+        delete_all_users()
```
nit: I think this could be `self.user.delete()` which might be slightly less work

czue commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,291 @@
+    @classmethod
+    def tearDownClass(cls):
+        for domain in Domain.get_all():
+            domain.delete()
```
i think this is a particularly slow function because it calls all sorts of other stuff. is there a reason this is necessary?

sravfeyn replied:
Was supposed to be temporary cleanup. Removed it

czue commented on custom/enikshay/tests/utils.py:
```diff
@@ -27,6 +27,115 @@
 from corehq.apps.users.models import CommCareUser
 
 
+def get_person_case_structure(case_id, user_id, extra_update={}):
```
nit: [mutable default args](http://docs.python-guide.org/en/latest/writing/gotchas/#mutable-default-arguments)

sravfeyn replied:
I am not updating `extra_update`, so it should be okay, no?

czue replied:
it should be, though is more future proof to make the default `None` and then have `extra_update = extra_update or {}` at the top of the function. just in case later someone else updates this function to modify it some how

sravfeyn replied:
Cool, updated

czue commented on custom/enikshay/tests/utils.py:
```diff
@@ -27,6 +27,115 @@
+        'secondary_contact_name_address': 'Mrs. Everestie',
+        'previous_tb_treatment': 'yes',
+        'nikshay_registered': "false",
+    }
```
didn't review these properties for correctness. let me know if you want me to

czue replied:
oh cool, see you were just refactoring this.

sravfeyn replied:
yeah

czue reviewed (commented):
this is looking great sravs. thanks for putting it all together and for the great tests.

wanted to start a conversation about a few possible scalability concerns I have. let me know if it's better to offline some of this. also would like @ctsims to sign off on the tests. also had a bunch of small stuff.

sravfeyn committed:
docstrings

sravfeyn committed:
better setup/teardown

sravfeyn committed:
try catch individual updates

sravfeyn committed:
two more props; new test-cases; refactor

sravfeyn committed:
mutable defaults

sravfeyn committed:
small refactor

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+            ),
+            {   # change
+                'aggregated_score_date_calculated': date(2016, 1, 20),
+                'expected_doses_taken': (10.0 / 7) * int(self.fixture_data['schedule1']),
```
I think this should be 0 here. Since there's no adherence data there's no expected doses. 

The "10" in this case should ~always match (adherence_latest_date_recorded - adherence_schedule_date_start)

sravfeyn replied:
Makes, sense. Updated

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+                []
+            ),
+            {   # change
+                'aggregated_score_date_calculated': date(2016, 1, 20),
```
I think this should match 'adherence_latest_date_recorded' from below ( no data, so it's 1 day before).

adherence_latest_date_recorded and aggregated_score_date_calculated will be the same if the purge date is after aggregated_score_date_calculated

sravfeyn replied:
Updated

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+                'expected_doses_taken': (5.0 / 7) * int(self.fixture_data['schedule1']),
+                'aggregated_score_count_taken': 1,
+                'adherence_latest_date_recorded': date(2016, 1, 15),
+                'adherence_total_doses_taken': 1
```
This looks right to me

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+            }
+        )
+
+    def test_adherence_date_greater_than_purge_date(self):
```
Great catch on this test case! Looks right to me, and is an important one to identify

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+                    (datetime(2016, 1, 11, 1), DOSE_TAKEN_INDICATORS[0]),
+                    (datetime(2016, 1, 11, 3), 'unknown'),
+                    (datetime(2016, 1, 12), DOSE_TAKEN_INDICATORS[0]),
+                    (datetime(2016, 1, 14), 'unknown'),
```
reminder that we want to use whatever we set up as the _missing_data thing instead of unknown, just for consistency with what will be coming in from the device.

sravfeyn replied:
Updated

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+                # ask clayton
+                'aggregated_score_count_taken': 2,
+                'adherence_latest_date_recorded': date(2016, 1, 21),
+                'adherence_total_doses_taken': 3
```
This should be 2, shouldn't it? 2 taken doses, 2 missed doses.

sravfeyn replied:
Yeah, updated

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+                'aggregated_score_date_calculated': date(2016, 1, 20),
+                'expected_doses_taken': (10.0 / 7) * int(self.fixture_data['schedule1']),
+                # ask clayton
+                'aggregated_score_count_taken': 2,
```
this is right

ctsims commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,363 @@
+            }
+        )
+
+    def test_two_doses_on_same_day_one_missed(self):
```
One other potential test case could be two doses on the same day, but different indicators (IE: one directly observed, one self reported), on the same day, which should also be recorded as 1 dose (not two)

sravfeyn replied:
Added this and few more test cases with various permutations of missed, unknown, less and greater dates.

sravfeyn committed:
change spec when no adherence-cases

sravfeyn committed:
string to const

sravfeyn committed:
don't known missed_dose; remove code

sravfeyn committed:
more test cases

sravfeyn committed:
comments

dimagimon commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,470 @@
+                'aggregated_score_count_taken': 0,
+                'adherence_total_doses_taken': 0
+                # 1 day before should be adherence_schedule_date_start
+                'adherence_latest_date_recorded': date(2016, 1, 9),
```
E901 SyntaxError: invalid syntax

sravfeyn committed:
comma; comment

sravfeyn committed:
try ES query for seach by reverse index

dimagimon commented on corehq/apps/reports/tests/test_esaccessors.py:
```diff
@@ -1029,6 +1029,33 @@ def _send_case_to_es(self,
+        child_case_id = uuid.uuid4().hex
+        parent_case_id = uuid.uuid4().hex
+        from casexml.apps.case.sharedmodels import CommCareCaseIndex
+        from casexml.apps.case.const import CASE_INDEX_CHILD, CASE_INDEX_EXTENSION
```
F401 'CASE_INDEX_CHILD' imported but unused

sravfeyn committed:
make indices nested type

sravfeyn committed:
test not-match is not matched

sravfeyn committed:
use elasticsearch to get cases; modify tests to ES

sravfeyn commented:
Hey @czue, I have implemented all of your feedback
- Switch to ES
- Remove redundant case lookups
- docstrings
- other assorted comments.

Can you review last few comments and merge if everything is ok?

I've also incorporated spec changes by Clayton, he is now signed off on the test cases (@ctsims).

sravfeyn commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,267 @@
+    def _get_open_episode_cases(self):
+        # return all open 'episode' cases
+        case_accessor = CaseAccessors(self.domain)
+        case_ids = case_accessor.get_open_case_ids_in_domain_by_type(CASE_TYPE_EPISODE)
```
Should I switch this also to ES? This is just one time call per task run. 

I wasn't able to find ES util for this, it's pretty easy to write, I will write one if it's needed

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,267 @@
+            if adherence_date > latest_date:
+                latest_date = adherence_date
+                latest_case = case
+        return latest_case
```
sorry i think i should have been clearer in my previous comment. so i think the main reason to push this to ES would be so that you can have the database return just the latest case and not have to grab every single one and do this check in python. unfortunately, since `adherence_date` is a dynamic property i don't think it's indexed which was why I was wondering if we could use the case search index or possibly make our own thing.

sravfeyn replied:
Ha, now I see the issue you are pointing out. There is potential custom ES queries that can be written here. I will dig into this a bit more and come up with a proposal

@NoahCarnahan @snopoke

proteusvacuum replied:
haven't followed this too closely, but the thing that @czue was mentioning is the `CaseSearchES`, which indexes each dynamic property already (it is turned on for this domain, so should be safe to use). You can see some examples [here](https://github.com/dimagi/commcare-hq/blob/8667a4159ad39d3307f92a3812d74006d610fd1b/corehq/apps/es/case_search.py#L38-L50)

proteusvacuum replied:
(you may have looked at that already though)

sravfeyn replied:
Yeah, this turned out to require more than `CaseSearchES`, currently looking into this issue

czue commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,267 @@
+            if adherence_value in DOSE_TAKEN_INDICATORS:
+                by_date[adherence_date].append(case)
+
+        return len(by_date.keys())
```
likewise i think maybe this aggregation could be pushed into ES (with the same caveats/complexity mentioned above).

czue reviewed (commented):
thanks for the updates Sravan! I'm still a bit worried about the processing times here if we're doing aggregations in memory in python. Since I'm not going to be around to push this conversation through to a final decision I wonder if it makes sense to loop in @NoahCarnahan or @snopoke and possibly even architecture team to try and figure out whether we think the current approach is sustainable? it is feeling risky to me

ctsims commented:
@sravfeyn In terms of what we move to ES: I think it would make a lot of sense for us to think through what we need to be indexing for this (and other) behaviors to be performant on eNikshay. 

Unfortunately I think this may be really tough to due to the complexities that have come up (IE: Needing to ignore "missing data" doses and needing to do "implicit de-aggregation" at a per-day basis) 

Would it be helpful for future work for us to articulate what, like, the "clean" data set is for adherence data that we'd end up running operations on or is it easier to figure out what's aggregatable task-by-task?

sravfeyn commented on corehq/pillows/mappings/case_mapping.py:
```diff
@@ -31,7 +31,7 @@
                                         'referenced_type': {'type': 'string'}
                                     },
-                                    'type': 'object'},
+                                    'type': 'nested'},
```
Yeah, I will add reindex label

emord replied:
This won't be reindexed because the `CASE_INDEX` did not change. @benrudolph this should get included with your PR

benrudolph replied:
👍 good catch

sravfeyn commented:
Clayton, definitely sounds like the final adherence data and future adherence calculations requirements will have an impact on how we solve ES/performance issues here. I will chat with you offline tomorrow re this

sravfeyn committed:
undo change

sravfeyn committed:
merge master

sravfeyn committed:
merge master

sravfeyn committed:
pass case-id

sravfeyn committed:
Add adherence ES UCR

sravfeyn committed:
another query

sravfeyn committed:
use UCR adherence queries in the task

dimagimon commented on custom/enikshay/tests/test_adherence_updates.py:
```diff
@@ -0,0 +1,498 @@
+from casexml.apps.case.mock import CaseFactory
+from corehq.apps.users.models import CommCareUser
+from corehq.apps.userreports.tasks import rebuild_indicators
+from corehq.form_processor.tests.utils import FormProcessorTestUtils
```
F811 redefinition of unused 'FormProcessorTestUtils' from line 7

sravfeyn committed:
Remove arg

sravfeyn commented:
@NoahCarnahan @proteusvacuum hey guys, can you review this, specifically the latest commits that push lot of python calculations to ES. I have one more spec update coming tomorrow, but wanted to have your review so that I can respond tomorrow. 

@emord buddy

proteusvacuum commented on corehq/apps/reports/tests/test_esaccessors.py:
```diff
@@ -1029,6 +1029,33 @@ def _send_case_to_es(self,
+             .run())
+
+        self.assertEqual(result.total, 1)
+        print result.hits
```
@sravfeyn this print statement is still here

proteusvacuum commented:
@sravfeyn I'm not sure I have enough context on this any more to do a proper review -- do you want someone to review alongside the spec? Or just to review moving to ES?

NoahCarnahan reviewed (commented):
Looks good Sraven. I'm curious if it's possible to use the existing cases or case search index for the ES queries? Didn't look too closely but I maybe they not store the indices in a searchable way?

emord commented on custom/enikshay/ucr/data_sources/adherence.json:
```diff
@@ -0,0 +1,101 @@
+{
+    "domains": [
```
Can you add the `server_environment` here

sravfeyn replied:
Added

emord commented on custom/enikshay/ucr/data_sources/adherence.json:
```diff
@@ -0,0 +1,101 @@
+    "config": {
+        "referenced_doc_type": "CommCareCase",
+        "engine_id": "ucr",
+        "backend_id": "ES",
```
@sravfeyn what are the benefits you see from using ES here? It looks like you're just doing some basic filtering for now.

emord commented on custom/enikshay/ucr/data_sources/adherence.json:
```diff
@@ -0,0 +1,101 @@
+    ],
+    "config": {
+        "referenced_doc_type": "CommCareCase",
+        "engine_id": "ucr",
```
You don't need `engine_id` for ES data sources

sravfeyn committed:
per test data

emord commented on corehq/apps/reports/analytics/esaccessors.py:
```diff
@@ -62,6 +63,17 @@ def get_active_case_counts_by_owner(domain, datespan, case_types=None, owner_ids
     return _get_case_case_counts_by_owner(domain, datespan, case_types, False, owner_ids)
 
 
+def get_reverse_indexed_cases_es(domain, case_id):
```
If this is not used I think it should be possible to remove the changes to the index and associated tests from this pr and move it to another pr if it will be needed later

sravfeyn replied:
Ha, makes sense, I will remove it if I don't need it. Still going back n forth on this

sravfeyn committed:
extra is-dose-taken conditions

sravfeyn committed:
merge master

sravfeyn committed:
server env

sravfeyn committed:
ucr

dimagimon commented on custom/enikshay/tests/utils.py:
```diff
@@ -212,7 +269,16 @@ def lab_referral(self):
-            self._get_adherence_case_structure(adherence_date, adherence_source, "unobserved_dose")
+            get_adherence_case_structure(
+                adherence_date.strftime('%Y-%m-%d')
+                self.episode_id,
```
E901 SyntaxError: invalid syntax

sravfeyn committed:
pep

sravfeyn committed:
comma

emord commented on custom/enikshay/ucr/data_sources/adherence.json:
```diff
@@ -0,0 +1,158 @@
+        "np-migration-3"
+    ],
+    "server_environment": [
+      "softlayer",
```
Trailing commas aren't valid in JSON. i'm not sure if python handles this correctly

sravfeyn committed:
merge conflicts

sravfeyn committed:
correct titles

sravfeyn committed:
trailing comma

sravfeyn committed:
reuse

sravfeyn committed:
memoize

sravfeyn committed:
case id

sravfeyn committed:
correct merge

sravfeyn committed:
remove extra util

sravfeyn committed:
remove reverse indexed case ES

sravfeyn committed:
move calc to Python

dimagimon commented on corehq/apps/reports/analytics/esaccessors.py:
```diff
@@ -1,6 +1,7 @@
 from collections import defaultdict, namedtuple
 from datetime import datetime
 
+from casexml.apps.case.models import CommCareCase
```
F401 'CommCareCase' imported but unused

sravfeyn committed:
correct merge

sravfeyn committed:
remove redundant test

sravfeyn committed:
comments

sravfeyn committed:
remove

sravfeyn committed:
split back

sravfeyn commented:
Offlined with Noah re performance issues, here is some of [the reasoning](https://docs.google.com/document/d/1_ARZHUrH0G-3wHGyiHA44sO8s2sDgi0B7ss4rqguVQY/edit) for current approach.

I will add datadog metrics to profile the task in a separate PR

NoahCarnahan commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,295 @@
+        if not lte:
+            return dose_taken_by_date.values().count(True)
+        else:
+            # any efficient way to do this - numpy, python bisect?
```
This doesn't look like it will be a bottleneck to me. Filtering this list is just `O(n)`.

sravfeyn replied:
👍

NoahCarnahan commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,295 @@
+                adherence_schedule_id = self.get_property('adherence_schedule_id') or DAILY_SCHEDULE_ID
+                doses_per_week = dose_data.get(adherence_schedule_id)
+                if doses_per_week:
+                    update['expected_doses_taken'] = ((
```
This will be a float, just want to confirm that's intended?

sravfeyn replied:
Hey @ctsims - I asked you this before. We decided on making this a floor right?

sravfeyn replied:
Offlined, made it to floor

NoahCarnahan commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,295 @@
+        """
+        self.episode = episode_case
+        self.case_updater = case_updater
+        self._cache_dose_taken_by_date = False
```
I don't think the value of this property ever changes.

sravfeyn replied:
Good catch! Fixed

NoahCarnahan commented on custom/enikshay/data_store.py:
```diff
@@ -0,0 +1,35 @@
+    def latest_adherence_date(self, episode_id):
+        result = self.dose_known_adherences(episode_id)
+        if len(result) > 0:
+            latest = sorted(result, key=lambda x: x['adherence_date'])[-1]
```
You're sorting in python because `dose_known_adherences` is memoized, right? Might just be worth adding a comment to this function to explain that that is why you aren't sorting in ES. I could imagine someone coming along later and thinking that they were speeding things up by moving the sort into ES.

NoahCarnahan commented on custom/enikshay/ucr/data_sources/adherence.json:
```diff
@@ -0,0 +1,158 @@
+        },
+        "table_id": "adherence",
+        "display_name": "adherence",
+        "configured_filter": {
```
I just realized that you could exclude the cases that don't have DOSE_KNOWN_INDICATORS from the data source itself. However, if you ever wanted to change this list, we would need to rebuild the data source.

NoahCarnahan reviewed (changes requested):
Looking good Sravan, just a couple small things.

sravfeyn committed:
sort in ES

sravfeyn committed:
Actually memoize

sravfeyn committed:
xor

sravfeyn commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,295 @@
+        else:
+            dose_taken_by_date = self._cache_dose_taken_by_date
+
+        if bool(lte) != bool(lte):
```
Oh yeah, fixed.

sravfeyn committed:
delete in teardown

NoahCarnahan commented on custom/enikshay/tasks.py:
```diff
@@ -0,0 +1,296 @@
+            dose_taken_by_date = self.calculate_doses_taken_by_day(adherence_cases)
+            self._cache_dose_taken_by_date = dose_taken_by_date
+        else:
+            dose_taken_by_date = self._cache_dose_taken_by_date
```
Hey, I just realized that you can probably do away with this memoization dict (which sort of obfuscates what's going on) by just passing `dose_taken_by_date` to `count_doses_taken()` instead of `adherence_cases`. I think this change would be nice to have, but not blocking the PR since I know this is on a tight timeline.

e.g.
```python
adherence_cases = self.get_valid_adherence_cases()
dose_taken_by_date = self.calculate_doses_taken_by_day(adherence_cases)
update["adherence_total_doses_taken"] = self.count_doses_taken(dose_taken_by_date)
# calculate 'aggregated_score_count_taken'
update["aggregated_score_count_taken"] = self.count_doses_taken(
	dose_taken_by_date,
	lte=adherence_schedule_date_start,
	gte=update["aggregated_score_date_calculated"]
)
```

sravfeyn replied:
Yeah, I was going back n forth between these two. I will actually change it to this. This is cleaner

NoahCarnahan commented:
(test failure looks real)

sravfeyn commented:
Yeah, I wasn't able to reproduce this test failure on local. I am gonna look in a bit.

sravfeyn committed:
fix tests?

sravfeyn committed:
more striaghtforward way of memoize

sravfeyn commented:
@NoahCarnahan ✅

sravfeyn committed:
floor of calc
